### ISSUE TITLE
Comments at end of line for polymorphic variants are duplicated on refmt.

### ISSUE THREAD

**Issue by jordwalke** [facebook/reason#1828]
Master branch:
```
  type widget_state = [
    | `DEFAULT /* here */
    | `HOVER
    | `ACTIVE
  ];
```

`refmt` leaves:

```
  type widget_state = [
    | `DEFAULT /* here */
    | `HOVER  /* here */
    | `ACTIVE  /* here */
  ];
```

And it causes the duplication to grow each refmt.

**Comment by emnh** [comment]
I was about to report this bug also, but then I saw it was already there 👍

**Comment by jordwalke** [comment]
@emnh did you only start seeing this in the master branch or 3.0.5? How are you testing that version? What kind of project?

**Comment by emnh** [comment]
I saw it on https://gist.github.com/Octachron/4e833a22844fd90cd6d15b0af927dab0 . My project is https://github.com/emnh/rts-reasonml and I use:

emh@rainmaker ~ 0% refmt --version                                                                                                                                                                                           (git)-[master]
Reason 3.0.4 @ 731e605

**Comment by jaredly** [comment]
Still an issue ([playground](https://reason.surge.sh/?syntax=Reason&code=ATAuE8AcFNgdwJYBMDm1QH0DOoCGpYBeYAbQCgQQAfYAAwBEBRAMQEEBVAGQBVgB6AFTAAFtABOsAXwqUatABIB5AGqMASjOp1WAYW4BJVTIC6AbiA&canvas=200))

**Comment by IwanKaramazow** [comment]
For future reference:

The problem is that the location of the specific row fields isn't tracked in the omp 4.04 ast we're currently using. Ocaml 4.06 does contain the location of the row fields. Due to the fact there's no location the comments think that they need to be attached to `type widget_state = [ ... ]`, which results in duplication. The comment interleaving algorithm doesn't know that it needs to attach them to the rows, no location to indicate so. Upgrading the parser to use 4.06 will fix this. (Not a small task)

**Comment by amiralies** [comment]
@anmonteiro is your branch going to be merged?

**Comment by glennsl** [comment]
This is fixed in 3.6.0, thankfully, because the problem increased exponentially with the number of comments, as shown in the diff below:

<details><summary>Insane refmt diff</summary><p>

```diff
   type pathCommand = [
-    // MoveTo
-    | `M(float, float) // x, y
-    | `m(float, float) // dx, dy
-    // LineTo
-    | `L(float, flaot) // x, y
-    | `l(float, float) // dx, dy
-    | `H(float) // x
-    | `h(flaot) // dx
-    | `V(float) // y
-    | `v(float) // dy
-    // CubicBezierCurve
-    | `C(float, float, float, float, float, float) // x1, y1, x2, y2, x, y
-    | `c(float, float, float, float, float, float) // dx1, dy1, dx2, dy2, dx, dy
-    | `S(float, float, float, float) // x2, y2, x, y
-    | `s(float, float, float, float) // dx2, dy2, dx, dy
-    // QuadraticBezierCurve
-    | `Q(float, float, float, float) // x1, y1, x, y
-    | `q(float, float, float, float) // dx1, dy1, dx, dy
-    | `T(float, float) // x, y
-    | `t(float, float) // dx, dy
-    // EllipticalArcCurve
-    | `A(float, float, [`large | `small], [`cw | `ccw], float, float) // rx, ry, large-arc-flag, sweep-flag, x, y
-    | `a(float, float, [`large | `small], [`cw | `ccw], float, float) // rx, ry, large-arc-flag, sweep-flag, dx, dy
-    // ClosePath
-    | `Z
-    | `z
+    | // MoveTo
+      `M(float, float) // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+      // LineTo
+      // CubicBezierCurve
+      // QuadraticBezierCurve
+      // EllipticalArcCurve
+      // ClosePath
+    | `m // x, y
+(float, float) // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `L // x, y
+ // dx, dy
+(float, flaot) // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `l // x, y
+ // dx, dy
+ // x, y
+(float, float) // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `H // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+(float) // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `h // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+(flaot) // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `V // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+(float) // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `v // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+(float) // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `C // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+(
+        float,
+        float,
+        float,
+        float,
+        float,
+        float,
+      ) // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `c // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+(
+        float,
+        float,
+        float,
+        float,
+        float,
+        float,
+      ) // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `S // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+(
+        float,
+        float,
+        float,
+        float,
+      ) // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `s // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+(
+        float,
+        float,
+        float,
+        float,
+      ) // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `Q // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+(
+        float,
+        float,
+        float,
+        float,
+      ) // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `q // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+(
+        float,
+        float,
+        float,
+        float,
+      ) // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `T // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+(
+        float,
+        float,
+      ) // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `t // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+(
+        float,
+        float,
+      ) // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `A // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+(
+        float,
+        float,
+        [ | `large | `small],
+        [ | `cw | `ccw],
+        float,
+        float,
+      ) // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `a // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+(
+        float,
+        float,
+        [ | `large | `small],
+        [ | `cw | `ccw],
+        float,
+        float,
+      ) // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `Z // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
+    | `z // x, y
+ // dx, dy
+ // x, y
+ // dx, dy
+ // x
+ // dx
+ // y
+ // dy
+ // x1, y1, x2, y2, x, y
+ // dx1, dy1, dx2, dy2, dx, dy
+ // x2, y2, x, y
+ // dx2, dy2, dx, dy
+ // x1, y1, x, y
+ // dx1, dy1, dx, dy
+ // x, y
+ // dx, dy
+ // rx, ry, large-arc-flag, sweep-flag, x, y
+ // rx, ry, large-arc-flag, sweep-flag, dx, dy
   ];
```
</p>
</details>
